### PR TITLE
[2.4.3] VecSim hybrid query - return empty iterator for invalid intersection child iterator

### PR DESCRIPTION
**Pull request by alonre24**
If the `NumExpected` method of the child iterator returns `UINT32_MAX`, it indicates that it is an invalid intersection iterator (if for example, one of the children is a phrase with a wildcard that has less than the minimal prefix, such as `@t:t*`).
 In that case, we return immediately an empty iterator instead of creating the hybrid iterator.

### CODE REVIEW

**Review by GuyAv46 — COMMENTED** [review]
Very nice and clean.
But is giving up completely the way other iterators behave when one/all of their child iterators are invalid? I thought that like in the parser, when child iterator is null, we behave like it is a wildcard, and in our case it means simple KNN query.

**Comment by alonre24** [comment]
> Very nice and clean. But is giving up completely the way other iterators behave when one/all of their child iterators are invalid? I thought that like in the parser, when child iterator is null, we behave like it is a wildcard, and in our case it means simple KNN query.

Well, I just tried to mimic the behaviour of RediSearch before the VSS addition, which is returning 0 results for invalid when one/all of their child iterators are invalid (and invalid wildcard in partucular). @ashtul right?

File: src/hybrid_reader.c [hunk 1]
```diff
@@ -340,12 +340,14 @@ IndexIterator *NewHybridVectorIterator(HybridIteratorParams hParams) {
...
   } else {
     // hi->searchMode is VECSIM_HYBRID_ADHOC_BF || VECSIM_HYBRID_BATCHES
-    hi->topResults = rm_malloc(heap_sizeof(hParams.query.k));
-    heap_init(hi->topResults, cmpVecSimResByScore, NULL, hParams.query.k);
-    hi->returnedResults = array_new(RSIndexResult *, hParams.query.k);
     // Get the estimated number of results that pass the child "sub-query filter". Note that
     // this is an upper bound, and might even be larger than the total vector index size.
     size_t subset_size = hParams.childIt->NumEstimated(hParams.childIt->ctx);
+    // UINT32_MAX is the default (invalid) value for indicating invalid intersection iterator.
+    if (subset_size == UINT32_MAX) {
```
Comment (DvirDukhan) [inline src/hybrid_reader.c:347]: if this is exposed (meaning, not iterator internal) have and variable/define with this value.

**Review by DvirDukhan — CHANGES_REQUESTED** [review]
very nice.
look at the comment

File: src/hybrid_reader.c [hunk 2]
```diff
@@ -340,12 +340,14 @@ IndexIterator *NewHybridVectorIterator(HybridIteratorParams hParams) {
     hi->searchMode = VECSIM_STANDARD_KNN;
   } else {
     // hi->searchMode is VECSIM_HYBRID_ADHOC_BF || VECSIM_HYBRID_BATCHES
-    hi->topResults = rm_malloc(heap_sizeof(hParams.query.k));
-    heap_init(hi->topResults, cmpVecSimResByScore, NULL, hParams.query.k);
-    hi->returnedResults = array_new(RSIndexResult *, hParams.query.k);
     // Get the estimated number of results that pass the child "sub-query filter". Note that
     // this is an upper bound, and might even be larger than the total vector index size.
     size_t subset_size = hParams.childIt->NumEstimated(hParams.childIt->ctx);
+    // UINT32_MAX is the default (invalid) value for indicating invalid intersection iterator.
```
Comment (ashtul) [inline src/hybrid_reader.c:346]: Why isn't it updated in `II_SortChildren`?

Comment (alonre24) [reply src/hybrid_reader.c:346]: Right... good catch thanks!

**Review by ashtul — DISMISSED** [review]

**Review by DvirDukhan — COMMENTED** [review]
nitpick

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**